pkg/pagerduty/pagerduty.go

```diff
@@ -126,6 +128,77 @@ func (p PagerDuty) AddNote(incidentID string, noteContent string) error {
 	return nil
 }
 
+// ExtractExternalIDFromCGHMAlertBody extracts from an Alert body until an external ID
+// the function is a member function but doesn't use any of the types
+func (_ PagerDuty) ExtractExternalIDFromCGHMAlertBody(data map[string]interface{}) (string, error) {
+
+	var err error
+
+	externalBody, err := marshalBodyToStruct(data)
+	if err != nil {
+		return "", fmt.Errorf("cannot marshal externalCHGMAlertBody: %w", err)
+	}
+
+	internalRawData := externalBody.Details.Notes
+	// check if the is a notes field, the default struct is provided if no match is found
+	if internalRawData == "" {
+		return "", AlertBodyDoesNotHaveNotesFieldErr{}
+	}
+
+	internalBody := internalCHGMAlertBody{}
+	err = yaml.Unmarshal([]byte(internalRawData), &internalBody)
+	if err != nil {
+		return "", NotesParseErr{Err: err}
+	}
+
+	// TODO: verify if no ClusterID is missing (doesn't break the test case, but should be noted
+
+	return internalBody.ClusterID, nil
+}
+
+// marshalBodyToStruct will marshal the map[string]interface{} to an externalCHGMAlertBody type
+func marshalBodyToStruct(body map[string]interface{}) (externalCHGMAlertBody, error) {
+	var ok bool
+	_, ok = body["details"]
+	if !ok {
+		return externalCHGMAlertBody{}, AlertBodyExternalParseErr{FailedProperty: ".details"}
+	}
+
+	details, ok := body["details"].(map[string]interface{})
+	if !ok {
+		return externalCHGMAlertBody{}, AlertBodyExternalCastErr{FailedProperty: ".details", ExpectedType: "map[string]interface{}", ActualType: reflect.TypeOf(body["details"]).String()}
+	}
+
+	notesInterface, ok := details["notes"]
+	if !ok {
+		return externalCHGMAlertBody{}, AlertBodyExternalParseErr{FailedProperty: ".details.notes"}
+	}
+
+	notes, ok := notesInterface.(string)
+	if !ok {
+		return externalCHGMAlertBody{}, AlertBodyExternalCastErr{FailedProperty: ".details.notes", ExpectedType: "string", ActualType: reflect.TypeOf(notes).String()}
+	}
+
+	res := externalCHGMAlertBody{}
+	res.Details.Notes = notes
+	return res, nil
```

Comment:
cleaner here if you have
`return externalCHGMALertBody{details: notes}, nil `

Reply:
Sure

Reply:
i had an anonymous struct in the `Details` field,  which made it hard to create the stuct there. so I removed the whole struct as I only need to extract the "notes" field